plugins/hydra_nevergrad_sweeper/hydra_plugins/hydra_nevergrad_sweeper/core.py

```diff
@@ -70,70 +165,74 @@ def make_parameter_from_commandline(string: str) -> Any:
            runtime. This provides full nevergrad flexibility at the cost of robustness.
            Eg.:"Log(a_min=0.001, a_max=0.1)"
          - anything else will be treated as a constant string
+       * a config definition dict for scalar parameters, with potential fields
+         init, lower, upper, step, log, integer
+       * a list for option parameters defined in config file
 
     Returns
     -------
     Parameter or str
         A Parameter if the string fitted one of the definitions, else the input string
     """
-    string = string.strip()
-    if "," in string:
-        return make_parameter_from_config(string.split(","))
-    if ":" in string:
-        *specs, b1, b2 = string.split(":")
-        a, b = [convert_to_deduced_type(x) for x in (b1, b2)]
-        assert isinstance(a, (int, float)), "Bounds must be scalars"
-        assert isinstance(b, (int, float)), "Bounds must be scalars"
-        description = {"lower": a, "upper": b, "log": "log" in specs}
-        if "int" in specs:
-            description["integer"] = True
-            if not all(isinstance(x, int) for x in (a, b)):
-                raise TypeError(
-                    f"Only integers should be provided for integer ranges (got {string})"
-                )
-            if b - a <= 6:
-                raise ValueError(
-                    "For integers with 6 or fewer values, use a choice instead"
-                )
-        return make_parameter_from_config(description)
-    return make_parameter_from_config(string)
-
-
-def make_parameter_from_config(description: Any) -> Any:
     # lazy initialization to avoid overhead when loading hydra
     import nevergrad as ng
 
-    # choice
+    # revert config parsing
 
     if isinstance(description, (ListConfig, list)):
-        choices = [convert_to_deduced_type(x) for x in description]
-        ordered = all(isinstance(c, (int, float)) for c in choices)
-        ordered &= all(
-            c0 <= c1 for c0, c1 in zip(choices[:-1], choices[1:])  # type: ignore
+        description = ",".join(description)
+    if isinstance(description, str):
+        # hacky nevergrad parameter
+        if description.startswith(tuple(dir(ng.p))):
+            param: ng.p.Parameter = eval(  # pylint: disable=eval-used
+                "ng.p." + description
+            )
+            assert isinstance(param, ng.p.Parameter)
+            return param
+        # cast to spec if possible
+        try:
+            description = CommandlineSpec.parse(description)
+        except ValueError:
+            pass
+    # convert scalar commandline specs to dict
+    if isinstance(description, CommandlineSpec) and description.bounds is not None:
+        description = ScalarConfigSpec(
+            lower=description.bounds[0],
+            upper=description.bounds[1],
+            log=description.log,
+            integer=description.cast == "int",
         )
```

Comment:
can you explain what is happening here?

you seem to be construvting the ScalarConfigSpec from multiple input types.
What are the cases?

is there a case when you use the one provided by Hydra as is?
(the Dictconfig that has the underlying ScalarConfigSpec type)?

Reply:
The docstring has details on this.
Basically this method converts any input to nevergrad parameter. That means:
- strings from commandline (which can actually be used in the config file as well) which are first read with `CommandlineSpec`, and then converted to `ScalarConfigSpec` if need be (i.e. if it's a range variable). This is what you pointed at
- list in the config files, which are considered as categorical variables
- dict in the config file, which are cast to `ScalarConfigSpec` for safety, then converted to the actual nevergrad type.

It's messy because there are several ways to define the parameters, and several type of parameters. I don't see simpler solutions though, so I tested it as much as possible.

considering `ScalarConfigSpec`, it's not used in the config, because it's not the only option to describe the variable (remember the `Union`?). It's basically only there to clarify the structure, and serves to check that there are no extra parameters which should not be there. Does that make more sense?

Reply:
Yes.
Where I would like us to converge is something like:

`hydra_opt_core`:
- defines common config structure and command line interface
- Implement conversion of the command line and the config into an object describing the variable (ScalarConfigSpec?), handling logic edge cases like specifying type in config but populating values from the line).

Now each plugin can call a single method, pass in it's config and command lines into a shared function and get back a list of ScalarConfigSpec it can use in it's logic.

How does that sound?
(This is not about this current diff, but more about where I would like us to go).

Reply:
I know it's been what we discussed about. I agree with it but that will take a while. I think the CommandlineParser is a small step in that direction (although it may need minor adaptation for identifying if cast was implicit or explicit).
However in the discussions we have overlooked the configuration file spec, which is probably critical to this end.

`ScalarConfigSpec` is only an intermediary structure: it wouldnt work in the config, since it's only dealing with scalars, not with choice variables.
Another aspect is that different sweeper may need different information to define a variable. For scalars I would like users to be able to define a step size, but I think this would not make sense for ax.
So should the config handle all possible cases and each plugins ignored irrelevant info? That could become messy, and may lead to incompatibilities
Or should the config file differentiate between the parametrization of each sweeper? That would lead to more robust spec, but would require a lot of duplications

Reply:
Maybe to summarize: we focused on a spec for the commandline so far. Commandline being a bit constrained, it has forced us to support only minimal features that all sweepers will probably share. On the other hand, config goes the other way round, so there are several options.

Reply:
Those are all things we can discuss as we start morphing the code in that direction.
specifically to your question about step size:
I think it would make sense for a plugin to just tell the user that step size is not supported if it's set and the plugin does not accept it.

Reply:
> I think it would make sense for a plugin to just tell the user that step size is not supported if it's set and the plugin does not accept it.

Then what is the point of having the same config if in practice we can't share it from one sweeper to another?

Reply:
1. Learning curve is shallower for switching between them.
2. Different sweepers will never be exactly the same, not in implementation and not in supported features set (and algorithms they can use). Having a unified config when possible will make switching between them easier.

A better decision might be to switch just issue a warning about the presence of the unsupported value and ignore it.